File: clients/java/src/main/java/com/mcp/client/UMLMCPClient.java
package com.mcp.client;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.Base64;
import java.util.UUID;

/**
 * Java client for UML MCP Server
 */
public class UMLMCPClient {

    private final String serverUrl;
    private final HttpClient httpClient;
    private final Gson gson;

    public UMLMCPClient(String serverUrl) {
        this.serverUrl = serverUrl.endsWith("/") ? serverUrl.substring(0, serverUrl.length() - 1) : serverUrl;
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
        this.gson = new GsonBuilder().setPrettyPrinting().create();
    }

    /**
     * Generate a UML diagram
     *
     * @param umlSource The PlantUML source code
     * @param format The output format (png, svg, txt)
     * @param diagramType The diagram type
     * @return MCPResponse object
     * @throws IOException if the request fails
     * @throws InterruptedException if the request is interrupted
     */
    public MCPResponse generateDiagram(String umlSource, String format, String diagramType)
        throws IOException, InterruptedException {

        MCPRequest request = new MCPRequest();
        request.setId(UUID.randomUUID().toString());
        request.setMethod("generate");

        MCPRequest.RequestParams params = new MCPRequest.RequestParams();
        params.setUml(umlSource);
        params.setFormat(format);
        params.setDiagramType(diagramType);
        request.setParams(params);

        String requestBody = gson.toJson(request);

        HttpRequest httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(serverUrl + "/mcp/generate"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(requestBody))
            .build();

        HttpResponse<String> response = httpClient.send(httpRequest,
            HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            throw new IOException("Server returned status code: " + response.statusCode() +
                ", body: " + response.body());
        }

        return gson.fromJson(response.body(), MCPResponse.class);
    }

    /**
     * Validate UML source code
     *
     * @param umlSource The PlantUML source code to validate
     * @return MCPResponse object
     * @throws IOException if the request fails
     * @throws InterruptedException if the request is interrupted
     */
    public MCPResponse validateUML(String umlSource) throws IOException, InterruptedException {
        MCPRequest request = new MCPRequest();
        request.setId(UUID.randomUUID().toString());
        request.setMethod("validate");

        MCPRequest.RequestParams params = new MCPRequest.RequestParams();
        params.setUml(umlSource);
        request.setParams(params);

        String requestBody = gson.toJson(request);

        HttpRequest httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(serverUrl + "/mcp/validate"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(requestBody))
            .build();

        HttpResponse<String> response = httpClient.send(httpRequest,
            HttpResponse.BodyHandlers.ofString());

        return gson.fromJson(response.body(), MCPResponse.class);
    }

    /**
     * Get server capabilities
     *
     * @return MCPResponse object
     * @throws IOException if the request fails
     * @throws InterruptedException if the request is interrupted
     */
    public MCPResponse getCapabilities() throws IOException, InterruptedException {
        MCPRequest request = new MCPRequest();
        request.setId(UUID.randomUUID().toString());
        request.setMethod("capabilities");

        String requestBody = gson.toJson(request);

        HttpRequest httpRequest = HttpRequest.newBuilder()
            .uri(URI.create(serverUrl + "/mcp/capabilities"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(requestBody))
            .build();

        HttpResponse<String> response = httpClient.send(httpRequest,
            HttpResponse.BodyHandlers.ofString());

        return gson.fromJson(response.body(), MCPResponse.class);
    }

    /**
     * Save diagram to file
     *
     * @param response The MCPResponse containing the diagram
     * @param outputPath The output file path
     * @throws IOException if saving fails
     */
    public void saveDiagram(MCPResponse response, String outputPath) throws IOException {
        if (response.getError() != null) {
            throw new IOException("Cannot save diagram: " + response.getError().getMessage());
        }

        MCPResponse.ResponseResult result = response.getResult();
        Path path = Paths.get(outputPath);

        if (result.getDiagramText() != null) {
            // Text format
            Files.writeString(path, result.getDiagramText());
        } else if (result.getDiagram() != null) {
            // Binary format (Base64 encoded)
            byte[] diagramBytes = Base64.getDecoder().decode(result.getDiagram());
            try (FileOutputStream fos = new FileOutputStream(outputPath)) {
                fos.write(diagramBytes);
            }
        } else {
            throw new IOException("No diagram data in response");
        }

        System.out.println("Diagram saved to: " + path.toAbsolutePath());
    }

    /**
     * Main method with example usage
     */
    public static void main(String[] args) {
        if (args.length < 1) {
            printUsage();
            System.exit(1);
        }

        String serverUrl = args.length > 1 ? args[1] : "http://localhost:8080";
        UMLMCPClient client = new UMLMCPClient(serverUrl);

        try {
            String command = args[0];

            switch (command) {
                case "generate":
                    generateExample(client, args);
                    break;
                case "validate":
                    validateExample(client, args);
                    break;
                case "capabilities":
                    capabilitiesExample(client);
                    break;
                default:
                    System.err.println("Unknown command: " + command);
                    printUsage();
                    System.exit(1);
            }

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void generateExample(UMLMCPClient client, String[] args)
        throws IOException, InterruptedException {

        String umlSource = args.length > 2 ? args[2] : getDefaultSequenceDiagram();
        String format = args.length > 3 ? args[3] : "png";
        String outputPath = args.length > 4 ? args[4] : "output." + format;

        System.out.println("Generating diagram...");
        System.out.println("Format: " + format);

        MCPResponse response = client.generateDiagram(umlSource, format, "sequence");

        if (response.getError() != null) {
            System.err.println("Error: " + response.getError().getMessage());
            System.exit(1);
        }

        client.saveDiagram(response, outputPath);
        System.out.println("Success!");
    }

    private static void validateExample(UMLMCPClient client, String[] args)
        throws IOException, InterruptedException {

        String umlSource = args.length > 2 ? args[2] : getDefaultSequenceDiagram();

        System.out.println("Validating UML...");
        MCPResponse response = client.validateUML(umlSource);

        if (response.getError() != null) {
            System.err.println("Error: " + response.getError().getMessage());
        } else {
            System.out.println("Result: " + response.getResult().getDiagramText());
        }
    }

    private static void capabilitiesExample(UMLMCPClient client)
        throws IOException, InterruptedException {

        System.out.println("Fetching server capabilities...");
        MCPResponse response = client.getCapabilities();

        if (response.getError() != null) {
            System.err.println("Error: " + response.getError().getMessage());
        } else {
            System.out.println(response.getResult().getDiagramText());
        }
    }

    private static String getDefaultSequenceDiagram() {
        return "@startuml\n" +
            "Alice -> Bob: Authentication Request\n" +
            "Bob --> Alice: Authentication Response\n" +
            "Alice -> Bob: Another authentication Request\n" +
            "Alice <-- Bob: Another authentication Response\n" +
            "@enduml";
    }

    private static void printUsage() {
        System.out.println("UML MCP Java Client");
        System.out.println("\nUsage: java -jar uml-mcp-java-client.jar <command> [server-url] [options]");
        System.out.println("\nCommands:");
        System.out.println("  generate [server-url] [uml-source] [format] [output-path]");
        System.out.println("  validate [server-url] [uml-source]");
        System.out.println("  capabilities [server-url]");
        System.out.println("\nExamples:");
        System.out.println("  java -jar client.jar generate http://localhost:8080");
        System.out.println("  java -jar client.jar generate http://localhost:8080 \"@startuml...@enduml\" png output.png");
        System.out.println("  java -jar client.jar capabilities http://localhost:8080");
    }

    // Inner classes for request/response (same as server models)
    private static class MCPRequest {
        private String jsonrpc = "2.0";
        private String id;
        private String method;
        private RequestParams params;

        public String getJsonrpc() { return jsonrpc; }
        public void setJsonrpc(String jsonrpc) { this.jsonrpc = jsonrpc; }
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }
        public String getMethod() { return method; }
        public void setMethod(String method) { this.method = method; }
        public RequestParams getParams() { return params; }
        public void setParams(RequestParams params) { this.params = params; }

        private static class RequestParams {
            private String uml;
            private String format;
            private String diagramType;

            public String getUml() { return uml; }
            public void setUml(String uml) { this.uml = uml; }
            public String getFormat() { return format; }
            public void setFormat(String format) { this.format = format; }
            public String getDiagramType() { return diagramType; }
            public void setDiagramType(String diagramType) { this.diagramType = diagramType; }
        }
    }

    private static class MCPResponse {
        private String jsonrpc;
        private String id;
        private ResponseResult result;
        private ResponseError error;

        public String getJsonrpc() { return jsonrpc; }
        public String getId() { return id; }
        public ResponseResult getResult() { return result; }
        public ResponseError getError() { return error; }

        private static class ResponseResult {
            private String diagram;
            private String diagramText;
            private String format;
            private String diagramType;
            private long timestamp;

            public String getDiagram() { return diagram; }
            public String getDiagramText() { return diagramText; }
            public String getFormat() { return format; }
            public String getDiagramType() { return diagramType; }
            public long getTimestamp() { return timestamp; }
        }

        private static class ResponseError {
            private int code;
            private String message;
            private String data;

            public int getCode() { return code; }
            public String getMessage() { return message; }
            public String getData() { return data; }
        }
    }
}

File: clients/python/uml_mcp_client.py
#!/usr/bin/env python3
"""
UML MCP Python Client

A Python client for interacting with the UML MCP Server to generate
UML diagrams using PlantUML.
"""

import argparse
import base64
import json
import sys
import uuid
from pathlib import Path
from typing import Optional, Dict, Any

try:
    import requests
except ImportError:
    print("Error: requests library is required. Install it with: pip install requests")
    sys.exit(1)


class MCPRequest:
    """MCP Request model"""

    def __init__(self, method: str, params: Dict[str, Any], request_id: Optional[str] = None):
        self.jsonrpc = "2.0"
        self.id = request_id or str(uuid.uuid4())
        self.method = method
        self.params = params

    def to_dict(self) -> Dict[str, Any]:
        return {
            "jsonrpc": self.jsonrpc,
            "id": self.id,
            "method": self.method,
            "params": self.params
        }


class MCPResponse:
    """MCP Response model"""

    def __init__(self, response_data: Dict[str, Any]):
        self.jsonrpc = response_data.get("jsonrpc")
        self.id = response_data.get("id")
        self.result = response_data.get("result")
        self.error = response_data.get("error")

    def has_error(self) -> bool:
        return self.error is not None

    def get_error_message(self) -> str:
        if self.error:
            return f"Error {self.error.get('code')}: {self.error.get('message')}"
        return ""


class UMLMCPClient:
    """
    Client for UML MCP Server
    """

    def __init__(self, server_url: str = "http://localhost:8080", timeout: int = 30):
        """
        Initialize the MCP client

        Args:
            server_url: The base URL of the MCP server
            timeout: Request timeout in seconds
        """
        self.server_url = server_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json'
        })

    def generate_diagram(
        self,
        uml_source: str,
        output_format: str = "png",
        diagram_type: str = "sequence"
    ) -> MCPResponse:
        """
        Generate a UML diagram

        Args:
            uml_source: PlantUML source code
            output_format: Output format (png, svg, txt)
            diagram_type: Diagram type (sequence, class, usecase, etc.)

        Returns:
            MCPResponse object

        Raises:
            requests.RequestException: If the request fails
        """
        params = {
            "uml": uml_source,
            "format": output_format,
            "diagramType": diagram_type
        }

        request = MCPRequest("generate", params)
        response = self._send_request("/mcp/generate", request)

        return MCPResponse(response)

    def validate_uml(self, uml_source: str) -> MCPResponse:
        """
        Validate PlantUML source code

        Args:
            uml_source: PlantUML source code to validate

        Returns:
            MCPResponse object

        Raises:
            requests.RequestException: If the request fails
        """
        params = {"uml": uml_source}
        request = MCPRequest("validate", params)
        response = self._send_request("/mcp/validate", request)

        return MCPResponse(response)

    def get_capabilities(self) -> MCPResponse:
        """
        Get server capabilities

        Returns:
            MCPResponse object

        Raises:
            requests.RequestException: If the request fails
        """
        request = MCPRequest("capabilities", {})
        response = self._send_request("/mcp/capabilities", request)

        return MCPResponse(response)

    def save_diagram(self, response: MCPResponse, output_path: str) -> None:
        """
        Save diagram to file

        Args:
            response: MCPResponse containing the diagram
            output_path: Path to save the diagram

        Raises:
            ValueError: If the response contains an error or no diagram data
            IOError: If saving fails
        """
        if response.has_error():
            raise ValueError(f"Cannot save diagram: {response.get_error_message()}")

        result = response.result
        if not result:
            raise ValueError("No result in response")

        path = Path(output_path)

        if result.get('diagramText'):
            # Text format
            path.write_text(result['diagramText'])
        elif result.get('diagram'):
            # Binary format (Base64 encoded)
            diagram_bytes = base64.b64decode(result['diagram'])
            path.write_bytes(diagram_bytes)
        else:
            raise ValueError("No diagram data in response")

        print(f"Diagram saved to: {path.absolute()}")

    def _send_request(self, endpoint: str, request: MCPRequest) -> Dict[str, Any]:
        """
        Send request to the server

        Args:
            endpoint: API endpoint
            request: MCPRequest object

        Returns:
            Response data as dictionary

        Raises:
            requests.RequestException: If the request fails
        """
        url = f"{self.server_url}{endpoint}"
        response = self.session.post(
            url,
            json=request.to_dict(),
            timeout=self.timeout
        )

        response.raise_for_status()
        return response.json()


def get_default_sequence_diagram() -> str:
    """Get a default sequence diagram example"""
    return """@startuml
Alice -> Bob: Authentication Request
Bob --> Alice: Authentication Response
Alice -> Bob: Another authentication Request
Alice <-- Bob: Another authentication Response
@enduml"""


def get_default_class_diagram() -> str:
    """Get a default class diagram example"""
    return """@startuml
class User {
    -String username
    -String password
    +login()
    +logout()
}

class Admin {
    +manageUsers()
}

User <|-- Admin
@enduml"""


def main():
    """Main entry point"""
    parser = argparse.ArgumentParser(
        description="UML MCP Python Client - Generate UML diagrams using PlantUML"
    )

    parser.add_argument(
        "command",
        choices=["generate", "validate", "capabilities"],
        help="Command to execute"
    )

    parser.add_argument(
        "--server",
        default="http://localhost:8080",
        help="MCP server URL (default: http://localhost:8080)"
    )

    parser.add_argument(
        "--uml",
        help="PlantUML source code or path to .puml file"
    )

    parser.add_argument(
        "--format",
        default="png",
        choices=["png", "svg", "txt"],
        help="Output format (default: png)"
    )

    parser.add_argument(
        "--output",
        help="Output file path (default: output.<format>)"
    )

    parser.add_argument(
        "--diagram-type",
        default="sequence",
        help="Diagram type (default: sequence)"
    )

    parser.add_argument(
        "--example",
        choices=["sequence", "class"],
        help="Use a built-in example diagram"
    )

    args = parser.parse_args()

    # Create client
    client = UMLMCPClient(server_url=args.server)

    try:
        if args.command == "generate":
            # Get UML source
            if args.example:
                if args.example == "sequence":
                    uml_source = get_default_sequence_diagram()
                else:
                    uml_source = get_default_class_diagram()
            elif args.uml:
                # Check if it's a file path
                uml_path = Path(args.uml)
                if uml_path.exists() and uml_path.is_file():
                    uml_source = uml_path.read_text()
                else:
                    uml_source = args.uml
            else:
                uml_source = get_default_sequence_diagram()
                print("No UML source provided, using default sequence diagram")

            # Generate diagram
            print(f"Generating {args.format} diagram...")
            response = client.generate_diagram(
                uml_source,
                args.format,
                args.diagram_type
            )

            if response.has_error():
                print(f"Error: {response.get_error_message()}", file=sys.stderr)
                sys.exit(1)

            # Save diagram
            output_path = args.output or f"output.{args.format}"
            client.save_diagram(response, output_path)
            print("Success!")

        elif args.command == "validate":
            if not args.uml and not args.example:
                print("Error: --uml or --example is required for validate command",
                      file=sys.stderr)
                sys.exit(1)

            # Get UML source
            if args.example:
                uml_source = (get_default_sequence_diagram() if args.example == "sequence"
                             else get_default_class_diagram())
            else:
                uml_path = Path(args.uml)
                if uml_path.exists() and uml_path.is_file():
                    uml_source = uml_path.read_text()
                else:
                    uml_source = args.uml

            print("Validating UML...")
            response = client.validate_uml(uml_source)

            if response.has_error():
                print(f"Error: {response.get_error_message()}", file=sys.stderr)
                sys.exit(1)

            print(f"Result: {response.result.get('diagramText')}")

        elif args.command == "capabilities":
            print("Fetching server capabilities...")
            response = client.get_capabilities()

            if response.has_error():
                print(f"Error: {response.get_error_message()}", file=sys.stderr)
                sys.exit(1)

            print(response.result.get('diagramText'))

    except requests.RequestException as e:
        print(f"Request failed: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
